anchore/syft
# feat: update spdx format model to produce valid spdx json documents

## Summary - add small adjustments so default spdx documents are valid

- Add Describes Relationship to spdx relationships
- Add DOCUMENT describes DOCUMENT relationship to final spdx format model
-  Add empty digests for files that are not regular files which cannot be linked back to their regular file

The above changes get syft to a point where it can produce SPDX documents that pass the different tooling validators found under [spdx-org](https://github.com/spdx?type=source). There are future enhancements we can make that allow syft to generate better digest information for symlinks and irregular files. 

### Future improvments
1) Symlinks can be updated in the SBOM to resolve to their real path and lookup the related digest built for that real path by the digest cataloger

2) The `DOCUMENT` `describes` relationship can be pointed to a specific package constructed by the SBOM. One proposal surrounding this is using the [primary package purpose field](https://spdx.github.io/spdx-spec/v2.3/package-information/#724-primary-package-purpose-field). For a container scan, we can use the `source` information to construct a package that references the scanned container with the primary package purpose of `CONTAINER`. The SPDX `DOCUMENT` `describes` relationship can then be pointed at that package.

#### Image showing new CLI tests for spdx validity against candidate images
<img width="1430" alt="Screenshot 2022-12-20 at 5 17 21 PM" src="https://user-images.githubusercontent.com/32073428/208777603-99f202bc-2438-4a98-9e85-15609876852f.png">

### To Test
```
SYFT_FILE_METADATA_CATALOGER_ENABLED=true SYFT_FILE_METADATA_DIGESTS=sha1 go run cmd/syft/main.go alpine:latest -o spdx > test.spdx
```
```
SYFT_FILE_METADATA_CATALOGER_ENABLED=true SYFT_FILE_METADATA_DIGESTS=sha1 go run cmd/syft/main.go alpine:latest -o spdx-json > test.json
```

^ The above outputs can be submitted to the SPDX validator:
https://tools.spdx.org/app/validate/

Note: To produce valid SPDX we still need to enable to digest cataloger because of the files analyzed field. Running this with no sha1 digest capability will result in `Found analyzed files for package xxx when analyzedFiles is set to false`

There is also automated testing now in syft to check against the tooling:
<img width="1155" alt="Screenshot 2022-12-20 at 11 12 19 PM" src="https://user-images.githubusercontent.com/32073428/208819685-7df5ea9d-52bb-425f-8931-af4f9fbc8b8e.png">

Signed-off-by: Christopher Phillips <christopher.phillips@anchore.com>
